ffi/src/connector/mod.rs

```diff
@@ -0,0 +1,177 @@
+pub mod config;
+pub mod result;
+
+#[diplomat::bridge]
+pub mod ffi {
+    use diplomat_runtime::DiplomatWriteable;
+    use ironrdp::connector::Sequence as _;
+    use std::fmt::Write;
+
+    use crate::{
+        error::{
+            ffi::{IronRdpError, IronRdpErrorKind},
+            NullPointerError,
+        },
+        utils::ffi::{SocketAddr, VecU8},
+    };
+
+    use super::config::ffi::Config;
+
+    #[diplomat::opaque] // We must use Option here, as ClientConnector is not Clone and have functions that consume it
+    pub struct ClientConnector(pub Option<ironrdp::connector::ClientConnector>);
+
+    #[diplomat::opaque]
+    pub struct ClientConnectorState(pub ironrdp::connector::ClientConnectorState);
+
+    #[diplomat::opaque]
+    pub struct ServerName(pub ironrdp::connector::ServerName);
+
+    // Basic Impl for ClientConnector
+    impl ClientConnector {
+        pub fn new(config: &Config) -> Box<ClientConnector> {
+            Box::new(ClientConnector(Some(ironrdp::connector::ClientConnector::new(
+                config.0.clone(),
+            ))))
+        }
+
+        /// Must use
+        pub fn with_server_addr(&mut self, server_addr: &SocketAddr) -> Result<(), Box<IronRdpError>> {
+            let Some(connector) = self.0.take() else {
+                return Err(IronRdpErrorKind::NullPointer.into());
+            };
+            let server_addr = server_addr.0;
+            self.0 = Some(connector.with_server_addr(server_addr));
+
+            Ok(())
+        }
+
+        // FIXME: We need to create opaque for ironrdp::svc::StaticChannelSet
+        /// Must use
+        pub fn with_static_channel_rdp_snd(&mut self) -> Result<(), Box<IronRdpError>> {
+            let Some(connector) = self.0.take() else {
+                return Err(IronRdpErrorKind::NullPointer.into());
+            };
+
+            self.0 = Some(connector.with_static_channel(ironrdp::rdpsnd::Rdpsnd::new()));
+
+            Ok(())
+        }
+
+        // FIXME: We need to create opaque for ironrdp::rdpdr::Rdpdr
+        /// Must use
+        pub fn with_static_channel_rdpdr(
+            &mut self,
+            computer_name: &str,
+            smart_card_device_id: u32,
+        ) -> Result<(), Box<IronRdpError>> {
+            let Some(connector) = self.0.take() else {
+                return Err(IronRdpErrorKind::NullPointer.into());
```

Comment:
naming: `IronRdpErrorKind::Consumed` instead of `NullPointer`

Reply:
updated